### PR TITLE
Fix/rounding changes

### PR DESCRIPTION
**Pull request by jcompagni10**
Change rounding to favor maker instead of taker. This is necessary to eliminate a large class of rounding-related errors and is a more optimal UX for market makers. 
As part of this change, we also enforce a trueLimitPrice to ensure taker LimitOrders are being executed at the intended price.

### CODE REVIEW

**Comment by pr0n00gler** [comment]
Please fix linter issues

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/8702335883

File: x/dex/keeper/liquidity.go [hunk 1]
```diff
@@ -130,15 +115,69 @@ func (k Keeper) SaveLiquidity(sdkCtx sdk.Context, liquidityI types.Liquidity) {
...
-	inAmount, outAmount math.Int,
-	liq types.Liquidity,
-) bool {
-	bookPrice := liq.Price()
-	truePrice := math_utils.NewPrecDecFromInt(outAmount).QuoInt(inAmount)
-	priceDiffFromExpected := truePrice.Sub(bookPrice)
-	pctDiff := priceDiffFromExpected.Quo(bookPrice)
-
-	return pctDiff.GT(maxTrueTakerSpread)
+func (k Keeper) TakerLimitOrderSwap(
```
Comment (NeverHappened) [inline x/dex/keeper/liquidity.go:120]: some comments describing takerLimitOrderSwap and MakerLimitOrder swap and why there is such distinction would be nice

File: x/dex/keeper/multihop_swap.go [hunk 2]
```diff
@@ -34,7 +34,7 @@ func (k Keeper) HopsToRouteData(
 		}
 		price, found := k.GetCurrPrice(ctx, tradePairID)
 		if !found {
-			return routeArr, types.ErrInsufficientLiquidity
+			return routeArr, types.ErrLimitPriceNotSatisfied
```
Comment (NeverHappened) [inline x/dex/keeper/multihop_swap.go:37]: maybe different error? it's kinda more about missing liquidity pools at all than limit price not satisfied. (yeah i know probably a rare occurence, but still)

File: x/dex/keeper/liquidity.go [hunk 3]
```diff
@@ -43,20 +42,6 @@ func (k Keeper) Swap(
 
 		inAmount, outAmount := liq.Swap(remainingTakerDenom, remainingMakerDenom)
 
-		// If (due to rounding) the actual price given to the maker is demonstrably unfair
-		// we do not save the results of the swap and we exit.
-		// While the decrease in price quality for the maker is semi-linear with the amount
-		// being swapped, it is possible that the next swap could yield a "fair" price.
-		// Nonetheless, once the remainingTakerDenom gets small enough to start causing unfair swaps
-		// it is much simpler to just abort.
-		if inAmount.IsZero() || isUnfairTruePrice(params.MaxTrueTakerSpread, inAmount, outAmount, liq) {
```
Comment (NeverHappened) [inline x/dex/keeper/liquidity.go:52]: now that we removed this check - i see you've added the check for truePrice < exitLimitPrice.
Do we have any tests for cases where this condition is not satisfied?

File: x/dex/keeper/core.go [hunk 4]
```diff
@@ -331,49 +331,38 @@ func (k Keeper) PlaceLimitOrderCore(
...
-		limitPrice, err = types.CalcPrice(tickIndexInToOut)
-		if err != nil {
-			return trancheKey, totalInCoin, swapInCoin, swapOutCoin, err
-		}
+	amountLeft := amountIn
+
+	// This is ok because tokenOut is provided to the constructor of PairID above
+	takerTradePairID := pairID.MustTradePairIDFromMaker(tokenOut)
+	var limitPrice math_utils.PrecDec
+	limitPrice, err = types.CalcPrice(tickIndexInToOut)
```
Comment (NeverHappened) [inline x/dex/keeper/core.go:339]: a little out of scope, but how about renaming tickIndexInToOut -> limitTickIndexInToOut? if I understand it correctly,  it's acting as a tick limit, not as a tick of an actual swap

**Review by NeverHappened — COMMENTED** [review]
nice!
I also suggest adding some tests from here https://github.com/neutron-org/neutron/pull/443/files
And we can add a test that checks multihop swap works on cheap to expensive routes

File: x/dex/keeper/multihop_swap.go (hunk 2 above)
Comment (jcompagni10) [reply x/dex/keeper/multihop_swap.go:37]: Decided to simplify things by having a single error here. From a user's perspective, if there's no liquidity or all of the liquidity is above their limitPrice, it's functionally the same.

File: x/dex/keeper/liquidity.go (hunk 3 above)
Comment (jcompagni10) [reply x/dex/keeper/liquidity.go:52]: https://github.com/neutron-org/neutron/blob/fix%2Frounding_changes/x/dex/keeper/integration_placelimitorder_test.go#L471
https://github.com/neutron-org/neutron/blob/fix%2Frounding_changes/x/dex/keeper/integration_placelimitorder_test.go#L304

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/8944816147

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9025929761

**Review by NicholasDotSol — DISMISSED** [review]

**Comment by pr0n00gler** [comment]
Please fix linter issues, otherwise lgtm

File: x/dex/keeper/liquidity.go [hunk 5]
```diff
@@ -130,15 +115,73 @@ func (k Keeper) SaveLiquidity(sdkCtx sdk.Context, liquidityI types.Liquidity) {
...
+		return sdk.Coin{}, sdk.Coin{}, types.ErrFoKLimitOrderNotFilled
+	}
+
+	if totalInCoin.Amount.IsZero() {
+		return sdk.Coin{}, sdk.Coin{}, types.ErrLimitPriceNotSatisfied
+	}
+
+	truePrice := math_utils.NewPrecDecFromInt(totalOutCoin.Amount).QuoInt(totalInCoin.Amount)
+
+	if truePrice.LT(limitPrice) {
```
Comment (NeverHappened) [inline x/dex/keeper/liquidity.go:149]: Is this correct, or should be the other way around?

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9085720921

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9276692068

**Review by NeverHappened — APPROVED** [review]

**Review by quasisamurai — APPROVED** [review]

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9217922297

File: x/dex/types/keys.go [hunk 6]
```diff
@@ -110,8 +110,13 @@ func LimitOrderTrancheUserAddressPrefix(address string) []byte {
 }
 
 func TimeBytes(timestamp time.Time) []byte {
-	unixMs := uint64(timestamp.UnixMilli())
-	str := utils.Uint64ToSortableString(unixMs)
+	var unixSecs uint64
+	// If timestamp is 0 use that instead of returning long negative number for unix time
+	if !timestamp.IsZero() {
```
Comment (swelf19) [inline x/dex/types/keys.go]: 1) i still not sure i understand what is it for
2) by changing timebytes representation, we change `LimitOrderExpirationKey`, do we need migration for existing orders?

**Review by pr0n00gler — APPROVED** [review]

**Review by swelf19 — APPROVED** [review]

**Review by NeverHappened — APPROVED** [review]

**Pull request merged**